uclibc backport libc mkostemp helpers

Backport upstream uClibc patches that add the mkostemp() family of
calls to stdlib:

commit 42d1b23fc0f3748b8bf474e456d6c44aa7e563fd
Author: Mike Frysinger <vapier@gentoo.org>
Date:   Wed Nov 14 00:30:54 2012 -0500

    libc/stdlib: add mkostemp helpers

    Some projects (like udev) are starting to use this.

    Imported from glibc.

    Signed-off-by: Mike Frysinger <vapier@gentoo.org>

commit 5084cf8b6b8f01890b62bfd0edf085d04887d084
Author: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
Date:   Sun Feb 10 17:21:15 2013 +0100

    stdlib: Fix fndecl of mkostemp, mkostemp64

    Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>

commit 638a23483b40c5b606ee323e6612e7e454e5154b
Author: Anthony G. Basile <blueness@gentoo.org>
Date:   Mon Oct 27 16:13:34 2014 -0400

    mkostemp: fix implementation

    mkostemp(char *template, int flags) generates a unique temporary
    filename from a template.  The flags parameter accepts three of
    the same flags as open(2): O_APPEND, O_CLOEXEC, and O_SYNC.  The
    current implementation of mkostemp(3) does not respect the flags
    and in fact confuses the flags with the file mode which should
    always be S_IRUSR | S_IWUSR.  This patch corrects this issue.

    Signed-off-by: Anthony G. Basile <blueness@gentoo.org>
    Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
---
 .../401-libc-stdlib-add-mkostemp-helpers.patch     |  155 +++++++++
 ...-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch |   44 +++
 .../0.9.32.1/403-mkostemp-fix-implementation.patch |  328 ++++++++++++++++++++
 .../401-libc-stdlib-add-mkostemp-helpers.patch     |  155 +++++++++
 ...-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch |   44 +++
 .../0.9.33.2/403-mkostemp-fix-implementation.patch |  328 ++++++++++++++++++++
 6 files changed, 1054 insertions(+)
 create mode 100644 patches/uClibc/0.9.32.1/401-libc-stdlib-add-mkostemp-helpers.patch
 create mode 100644 patches/uClibc/0.9.32.1/402-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch
 create mode 100644 patches/uClibc/0.9.32.1/403-mkostemp-fix-implementation.patch
 create mode 100644 patches/uClibc/0.9.33.2/401-libc-stdlib-add-mkostemp-helpers.patch
 create mode 100644 patches/uClibc/0.9.33.2/402-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch
 create mode 100644 patches/uClibc/0.9.33.2/403-mkostemp-fix-implementation.patch

diff --git a/patches/uClibc/0.9.32.1/401-libc-stdlib-add-mkostemp-helpers.patch b/patches/uClibc/0.9.32.1/401-libc-stdlib-add-mkostemp-helpers.patch
new file mode 100644
index 0000000..1a8d7a5
--- /dev/null
+++ b/patches/uClibc/0.9.32.1/401-libc-stdlib-add-mkostemp-helpers.patch
@@ -0,0 +1,155 @@
+From 42d1b23fc0f3748b8bf474e456d6c44aa7e563fd Mon Sep 17 00:00:00 2001
+From: Mike Frysinger <vapier@gentoo.org>
+Date: Wed, 14 Nov 2012 00:30:54 -0500
+Subject: [PATCH 1/1] libc/stdlib: add mkostemp helpers
+
+Some projects (like udev) are starting to use this.
+
+Imported from glibc.
+
+Signed-off-by: Mike Frysinger <vapier@gentoo.org>
+---
+ include/stdlib.h         | 23 +++++++++++++++++++++++
+ libc/stdlib/Makefile.in  |  4 ++--
+ libc/stdlib/mkostemp.c   | 32 ++++++++++++++++++++++++++++++++
+ libc/stdlib/mkostemp64.c | 33 +++++++++++++++++++++++++++++++++
+ 4 files changed, 90 insertions(+), 2 deletions(-)
+ create mode 100644 libc/stdlib/mkostemp.c
+ create mode 100644 libc/stdlib/mkostemp64.c
+
+diff --git a/include/stdlib.h b/include/stdlib.h
+index 354fc66..79ccc55 100644
+--- a/include/stdlib.h
++++ b/include/stdlib.h
+@@ -652,6 +652,29 @@ extern int mkstemp64 (char *__template) __nonnull ((1)) __wur;
+ extern char *mkdtemp (char *__template) __THROW __nonnull ((1)) __wur;
+ #endif
+ 
++#ifdef __USE_GNU
++/* Generate a unique temporary file name from TEMPLATE similar to
++   mkstemp.  But allow the caller to pass additional flags which are
++   used in the open call to create the file..
++
++   This function is a possible cancellation point and therefore not
++   marked with __THROW.  */
++# ifndef __USE_FILE_OFFSET64
++extern int mkostemp (char *__template, int __flags) __nonnull ((1)) __wur;
++# else
++#  ifdef __REDIRECT
++extern int __REDIRECT (mkostemp, (char *__template, int __flags), mkostemp64)
++     __nonnull ((1)) __wur;
++#  else
++#   define mkostemp mkostemp64
++#  endif
++# endif
++# ifdef __USE_LARGEFILE64
++extern int mkostemp64 (char *__template, int __flags) __nonnull ((1)) __wur;
++# endif
++
++#endif
++
+ 
+ __BEGIN_NAMESPACE_STD
+ /* Execute the given line as a shell command.
+diff --git a/libc/stdlib/Makefile.in b/libc/stdlib/Makefile.in
+index 3166b8e..b92f7ce 100644
+--- a/libc/stdlib/Makefile.in
++++ b/libc/stdlib/Makefile.in
+@@ -12,7 +12,7 @@ include $(top_srcdir)libc/stdlib/malloc-simple/Makefile.in
+ include $(top_srcdir)libc/stdlib/malloc-standard/Makefile.in
+ 
+ CSRC-y := \
+-	abort.c getenv.c mkdtemp.c realpath.c canonicalize.c mkstemp.c \
++	abort.c getenv.c mkdtemp.c realpath.c canonicalize.c mkstemp.c mkostemp.c \
+ 	rand.c random.c random_r.c setenv.c div.c ldiv.c lldiv.c \
+ 	getpt.c drand48-iter.c jrand48.c \
+ 	jrand48_r.c lcong48.c lrand48.c lrand48_r.c mrand48.c mrand48_r.c nrand48.c \
+@@ -21,7 +21,7 @@ CSRC-y := \
+ CSRC-$(UCLIBC_HAS_ADVANCED_REALTIME) += posix_memalign.c
+ CSRC-$(UCLIBC_HAS_PTY) += grantpt.c unlockpt.c ptsname.c
+ CSRC-$(UCLIBC_HAS_ARC4RANDOM) += arc4random.c
+-CSRC-$(UCLIBC_HAS_LFS) += mkstemp64.c
++CSRC-$(UCLIBC_HAS_LFS) += mkstemp64.c mkostemp64.c
+ CSRC-$(UCLIBC_HAS_FLOATS) += drand48.c drand48_r.c erand48.c erand48_r.c
+ CSRC-$(if $(findstring yy,$(UCLIBC_HAS_FLOATS)$(UCLIBC_SUSV3_LEGACY)),y) += \
+ 	gcvt.c
+diff --git a/libc/stdlib/mkostemp.c b/libc/stdlib/mkostemp.c
+new file mode 100644
+index 0000000..93b50fc
+--- /dev/null
++++ b/libc/stdlib/mkostemp.c
+@@ -0,0 +1,32 @@
++/* Copyright (C) 1998-2012 Free Software Foundation, Inc.
++   This file is part of the GNU C Library.
++
++   The GNU C Library is free software; you can redistribute it and/or
++   modify it under the terms of the GNU Lesser General Public
++   License as published by the Free Software Foundation; either
++   version 2.1 of the License, or (at your option) any later version.
++
++   The GNU C Library is distributed in the hope that it will be useful,
++   but WITHOUT ANY WARRANTY; without even the implied warranty of
++   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
++   Lesser General Public License for more details.
++
++   You should have received a copy of the GNU Lesser General Public
++   License along with the GNU C Library; if not, see
++   <http://www.gnu.org/licenses/>.  */
++
++#include <stdio.h>
++#include <stdlib.h>
++#include "../misc/internals/tempname.h"
++
++/* Generate a unique temporary file name from TEMPLATE.
++   The last six characters of TEMPLATE must be "XXXXXX";
++   they are replaced with a string that makes the filename unique.
++   Then open the file and return a fd. */
++int
++mkostemp (template, flags)
++     char *template;
++     int flags;
++{
++  return __gen_tempname (template, __GT_FILE, flags);
++}
+diff --git a/libc/stdlib/mkostemp64.c b/libc/stdlib/mkostemp64.c
+new file mode 100644
+index 0000000..5509d8c
+--- /dev/null
++++ b/libc/stdlib/mkostemp64.c
+@@ -0,0 +1,33 @@
++/* Copyright (C) 2000-2012 Free Software Foundation, Inc.
++   This file is part of the GNU C Library.
++
++   The GNU C Library is free software; you can redistribute it and/or
++   modify it under the terms of the GNU Lesser General Public
++   License as published by the Free Software Foundation; either
++   version 2.1 of the License, or (at your option) any later version.
++
++   The GNU C Library is distributed in the hope that it will be useful,
++   but WITHOUT ANY WARRANTY; without even the implied warranty of
++   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
++   Lesser General Public License for more details.
++
++   You should have received a copy of the GNU Lesser General Public
++   License along with the GNU C Library; if not, see
++   <http://www.gnu.org/licenses/>.  */
++
++#include <fcntl.h>
++#include <stdio.h>
++#include <stdlib.h>
++#include "../misc/internals/tempname.h"
++
++/* Generate a unique temporary file name from TEMPLATE.
++   The last six characters of TEMPLATE must be "XXXXXX";
++   they are replaced with a string that makes the filename unique.
++   Then open the file and return a fd. */
++int
++mkostemp64 (template, flags)
++     char *template;
++     int flags;
++{
++  return __gen_tempname (template, __GT_BIGFILE, flags | O_LARGEFILE);
++}
+-- 
+1.9.1
+
diff --git a/patches/uClibc/0.9.32.1/402-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch b/patches/uClibc/0.9.32.1/402-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch
new file mode 100644
index 0000000..b086be2
--- /dev/null
+++ b/patches/uClibc/0.9.32.1/402-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch
@@ -0,0 +1,44 @@
+From 5084cf8b6b8f01890b62bfd0edf085d04887d084 Mon Sep 17 00:00:00 2001
+From: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
+Date: Sun, 10 Feb 2013 17:21:15 +0100
+Subject: [PATCH 1/1] stdlib: Fix fndecl of mkostemp, mkostemp64
+
+Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
+---
+ libc/stdlib/mkostemp.c   | 4 +---
+ libc/stdlib/mkostemp64.c | 4 +---
+ 2 files changed, 2 insertions(+), 6 deletions(-)
+
+diff --git a/libc/stdlib/mkostemp.c b/libc/stdlib/mkostemp.c
+index 93b50fc..0369235 100644
+--- a/libc/stdlib/mkostemp.c
++++ b/libc/stdlib/mkostemp.c
+@@ -24,9 +24,7 @@
+    they are replaced with a string that makes the filename unique.
+    Then open the file and return a fd. */
+ int
+-mkostemp (template, flags)
+-     char *template;
+-     int flags;
++mkostemp (char *template, int flags)
+ {
+   return __gen_tempname (template, __GT_FILE, flags);
+ }
+diff --git a/libc/stdlib/mkostemp64.c b/libc/stdlib/mkostemp64.c
+index 5509d8c..d21def5 100644
+--- a/libc/stdlib/mkostemp64.c
++++ b/libc/stdlib/mkostemp64.c
+@@ -25,9 +25,7 @@
+    they are replaced with a string that makes the filename unique.
+    Then open the file and return a fd. */
+ int
+-mkostemp64 (template, flags)
+-     char *template;
+-     int flags;
++mkostemp64 (char *template, int flags)
+ {
+   return __gen_tempname (template, __GT_BIGFILE, flags | O_LARGEFILE);
+ }
+-- 
+1.9.1
+
diff --git a/patches/uClibc/0.9.32.1/403-mkostemp-fix-implementation.patch b/patches/uClibc/0.9.32.1/403-mkostemp-fix-implementation.patch
new file mode 100644
index 0000000..bb83d73
--- /dev/null
+++ b/patches/uClibc/0.9.32.1/403-mkostemp-fix-implementation.patch
@@ -0,0 +1,328 @@
+From 638a23483b40c5b606ee323e6612e7e454e5154b Mon Sep 17 00:00:00 2001
+From: "Anthony G. Basile" <blueness@gentoo.org>
+Date: Mon, 27 Oct 2014 16:13:34 -0400
+Subject: [PATCH 1/1] mkostemp: fix implementation
+
+mkostemp(char *template, int flags) generates a unique temporary
+filename from a template.  The flags parameter accepts three of
+the same flags as open(2): O_APPEND, O_CLOEXEC, and O_SYNC.  The
+current implementation of mkostemp(3) does not respect the flags
+and in fact confuses the flags with the file mode which should
+always be S_IRUSR | S_IWUSR.  This patch corrects this issue.
+
+Signed-off-by: Anthony G. Basile <blueness@gentoo.org>
+Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
+---
+ libc/inet/getaddrinfo.c               |  2 +-
+ libc/misc/internals/tempname.c        |  6 ++---
+ libc/misc/internals/tempname.h        |  2 +-
+ libc/stdio/tempnam.c                  |  2 +-
+ libc/stdio/tmpfile.c                  |  2 +-
+ libc/stdio/tmpnam.c                   |  2 +-
+ libc/stdio/tmpnam_r.c                 |  2 +-
+ libc/stdlib/mkdtemp.c                 |  2 +-
+ libc/stdlib/mkostemp.c                |  4 +++-
+ libc/stdlib/mkostemp64.c              |  2 +-
+ libc/stdlib/mkstemp.c                 |  2 +-
+ libc/stdlib/mkstemp64.c               |  2 +-
+ libc/stdlib/mktemp.c                  |  2 +-
+ libpthread/nptl/sem_open.c            |  2 +-
+ test/.gitignore                       |  2 ++
+ test/stdlib/test-mkostemp-O_CLOEXEC.c | 45 +++++++++++++++++++++++++++++++++++
+ test/stdlib/test-mkostemp-child.c     | 22 +++++++++++++++++
+ 17 files changed, 87 insertions(+), 16 deletions(-)
+ create mode 100644 test/stdlib/test-mkostemp-O_CLOEXEC.c
+ create mode 100644 test/stdlib/test-mkostemp-child.c
+
+diff --git a/libc/inet/getaddrinfo.c b/libc/inet/getaddrinfo.c
+index b61d69c..168adb1 100644
+--- a/libc/inet/getaddrinfo.c
++++ b/libc/inet/getaddrinfo.c
+@@ -308,7 +308,7 @@ gaih_local(const char *name, const struct gaih_service *service,
+ 		char *buf = ((struct sockaddr_un *)ai->ai_addr)->sun_path;
+ 
+ 		if (__path_search(buf, L_tmpnam, NULL, NULL, 0) != 0
+-		 || __gen_tempname(buf, __GT_NOCREATE, 0) != 0
++		 || __gen_tempname(buf, __GT_NOCREATE, 0, 0) != 0
+ 		) {
+ 			return -EAI_SYSTEM;
+ 		}
+diff --git a/libc/misc/internals/tempname.c b/libc/misc/internals/tempname.c
+index 18fd823..edcc31c 100644
+--- a/libc/misc/internals/tempname.c
++++ b/libc/misc/internals/tempname.c
+@@ -177,7 +177,7 @@ static void brain_damaged_fillrand(unsigned char *buf, unsigned int len)
+    __GT_DIR:            create a directory with given mode.
+ 
+ */
+-int __gen_tempname (char *tmpl, int kind, mode_t mode)
++int __gen_tempname (char *tmpl, int kind, int flags, mode_t mode)
+ {
+     char *XXXXXX;
+     unsigned int i;
+@@ -219,11 +219,11 @@ int __gen_tempname (char *tmpl, int kind, mode_t mode)
+ 			fd = 0;
+ 		}
+ 	    case __GT_FILE:
+-		fd = open (tmpl, O_RDWR | O_CREAT | O_EXCL, mode);
++		fd = open (tmpl, O_RDWR | O_CREAT | O_EXCL | flags, mode);
+ 		break;
+ #if defined __UCLIBC_HAS_LFS__
+ 	    case __GT_BIGFILE:
+-		fd = open64 (tmpl, O_RDWR | O_CREAT | O_EXCL, mode);
++		fd = open64 (tmpl, O_RDWR | O_CREAT | O_EXCL | flags, mode);
+ 		break;
+ #endif
+ 	    case __GT_DIR:
+diff --git a/libc/misc/internals/tempname.h b/libc/misc/internals/tempname.h
+index e75b632..edfe26d 100644
+--- a/libc/misc/internals/tempname.h
++++ b/libc/misc/internals/tempname.h
+@@ -10,7 +10,7 @@ extern int ___path_search (char *tmpl, size_t tmpl_len, const char *dir,
+ 	        const char *pfx /*, int try_tmpdir */) attribute_hidden;
+ #define __path_search(tmpl, tmpl_len, dir, pfx, try_tmpdir) ___path_search(tmpl, tmpl_len, dir, pfx)
+ 
+-extern int __gen_tempname (char *__tmpl, int __kind, mode_t mode);
++extern int __gen_tempname (char *__tmpl, int __kind, int flags, mode_t mode);
+ 
+ /* The __kind argument to __gen_tempname may be one of: */
+ #define __GT_FILE     0       /* create a file */
+diff --git a/libc/stdio/tempnam.c b/libc/stdio/tempnam.c
+index 232ed02..74bb26e 100644
+--- a/libc/stdio/tempnam.c
++++ b/libc/stdio/tempnam.c
+@@ -35,7 +35,7 @@ tempnam (const char *dir, const char *pfx)
+   if (__path_search (buf, FILENAME_MAX, dir, pfx, 1))
+     return NULL;
+ 
+-  if (__gen_tempname (buf, __GT_NOCREATE, 0))
++  if (__gen_tempname (buf, __GT_NOCREATE, 0, 0))
+     return NULL;
+ 
+   return strdup (buf);
+diff --git a/libc/stdio/tmpfile.c b/libc/stdio/tmpfile.c
+index a9bf474..83c85b5 100644
+--- a/libc/stdio/tmpfile.c
++++ b/libc/stdio/tmpfile.c
+@@ -35,7 +35,7 @@ FILE * tmpfile (void)
+ 
+     if (__path_search (buf, FILENAME_MAX, NULL, "tmpf", 0))
+ 	return NULL;
+-    fd = __gen_tempname (buf, __GT_FILE, S_IRUSR | S_IWUSR);
++    fd = __gen_tempname (buf, __GT_FILE, 0, S_IRUSR | S_IWUSR);
+     if (fd < 0)
+ 	return NULL;
+ 
+diff --git a/libc/stdio/tmpnam.c b/libc/stdio/tmpnam.c
+index 88b9bff..ffed862 100644
+--- a/libc/stdio/tmpnam.c
++++ b/libc/stdio/tmpnam.c
+@@ -40,7 +40,7 @@ tmpnam (char *s)
+ 			0))
+     return NULL;
+ 
+-  if (__builtin_expect (__gen_tempname (tmpbuf, __GT_NOCREATE, 0), 0))
++  if (__builtin_expect (__gen_tempname (tmpbuf, __GT_NOCREATE, 0, 0), 0))
+     return NULL;
+ 
+   if (s == NULL)
+diff --git a/libc/stdio/tmpnam_r.c b/libc/stdio/tmpnam_r.c
+index 8cde84d..bfd60a4 100644
+--- a/libc/stdio/tmpnam_r.c
++++ b/libc/stdio/tmpnam_r.c
+@@ -27,7 +27,7 @@ char * tmpnam_r (char *s)
+ 
+     if (__path_search (s, L_tmpnam, NULL, NULL, 0))
+ 	return NULL;
+-    if (__gen_tempname (s, __GT_NOCREATE, 0))
++    if (__gen_tempname (s, __GT_NOCREATE, 0, 0))
+ 	return NULL;
+ 
+     return s;
+diff --git a/libc/stdlib/mkdtemp.c b/libc/stdlib/mkdtemp.c
+index da7598a..e6d4a36 100644
+--- a/libc/stdlib/mkdtemp.c
++++ b/libc/stdlib/mkdtemp.c
+@@ -29,7 +29,7 @@
+    (This function comes from OpenBSD.) */
+ char * mkdtemp (char *template)
+ {
+-  if (__gen_tempname (template, __GT_DIR, S_IRUSR | S_IWUSR | S_IXUSR))
++  if (__gen_tempname (template, __GT_DIR, 0, S_IRUSR | S_IWUSR | S_IXUSR))
+     return NULL;
+   else
+     return template;
+diff --git a/libc/stdlib/mkostemp.c b/libc/stdlib/mkostemp.c
+index 0369235..912be30 100644
+--- a/libc/stdlib/mkostemp.c
++++ b/libc/stdlib/mkostemp.c
+@@ -17,6 +17,7 @@
+ 
+ #include <stdio.h>
+ #include <stdlib.h>
++#include <fcntl.h>
+ #include "../misc/internals/tempname.h"
+ 
+ /* Generate a unique temporary file name from TEMPLATE.
+@@ -26,5 +27,6 @@
+ int
+ mkostemp (char *template, int flags)
+ {
+-  return __gen_tempname (template, __GT_FILE, flags);
++  flags -= flags & O_ACCMODE; /* Remove O_RDONLY, O_WRONLY, and O_RDWR. */
++  return __gen_tempname (template, __GT_FILE, flags, S_IRUSR | S_IWUSR);
+ }
+diff --git a/libc/stdlib/mkostemp64.c b/libc/stdlib/mkostemp64.c
+index d21def5..c6d6d84 100644
+--- a/libc/stdlib/mkostemp64.c
++++ b/libc/stdlib/mkostemp64.c
+@@ -27,5 +27,5 @@
+ int
+ mkostemp64 (char *template, int flags)
+ {
+-  return __gen_tempname (template, __GT_BIGFILE, flags | O_LARGEFILE);
++  return __gen_tempname (template, __GT_BIGFILE, flags | O_LARGEFILE, S_IRUSR | S_IWUSR | S_IXUSR);
+ }
+diff --git a/libc/stdlib/mkstemp.c b/libc/stdlib/mkstemp.c
+index 61c7175..a3a1595 100644
+--- a/libc/stdlib/mkstemp.c
++++ b/libc/stdlib/mkstemp.c
+@@ -26,5 +26,5 @@
+    Then open the file and return a fd. */
+ int mkstemp (char *template)
+ {
+-    return __gen_tempname (template, __GT_FILE, S_IRUSR | S_IWUSR);
++    return __gen_tempname (template, __GT_FILE, 0, S_IRUSR | S_IWUSR);
+ }
+diff --git a/libc/stdlib/mkstemp64.c b/libc/stdlib/mkstemp64.c
+index e29be2d..6f2ee3e 100644
+--- a/libc/stdlib/mkstemp64.c
++++ b/libc/stdlib/mkstemp64.c
+@@ -26,5 +26,5 @@
+    Then open the file and return a fd. */
+ int mkstemp64 (char *template)
+ {
+-    return __gen_tempname (template, __GT_BIGFILE, S_IRUSR | S_IWUSR);
++    return __gen_tempname (template, __GT_BIGFILE, 0, S_IRUSR | S_IWUSR);
+ }
+diff --git a/libc/stdlib/mktemp.c b/libc/stdlib/mktemp.c
+index edd001d..1ff93da 100644
+--- a/libc/stdlib/mktemp.c
++++ b/libc/stdlib/mktemp.c
+@@ -24,7 +24,7 @@
+  * they are replaced with a string that makes the filename unique.  */
+ char *mktemp(char *template)
+ {
+-	if (__gen_tempname (template, __GT_NOCREATE, 0) < 0)
++	if (__gen_tempname (template, __GT_NOCREATE, 0, 0) < 0)
+ 		/* We return the null string if we can't find a unique file name.  */
+ 		template[0] = '\0';
+ 
+diff --git a/libpthread/nptl/sem_open.c b/libpthread/nptl/sem_open.c
+index 1b36164..3a72079 100644
+--- a/libpthread/nptl/sem_open.c
++++ b/libpthread/nptl/sem_open.c
+@@ -336,7 +336,7 @@ sem_open (const char *name, int oflag, ...)
+       mempcpy (mempcpy (tmpfname, mountpoint.dir, mountpoint.dirlen),
+ 	"XXXXXX", 7);
+ 
+-      fd = __gen_tempname (tmpfname, __GT_FILE, mode);
++      fd = __gen_tempname (tmpfname, __GT_FILE, 0, mode);
+       if (fd == -1)
+         return SEM_FAILED;
+ 
+diff --git a/test/.gitignore b/test/.gitignore
+index 8f32031..5944f0a 100644
+--- a/test/.gitignore
++++ b/test/.gitignore
+@@ -274,6 +274,8 @@ stdlib/testarc4random
+ stdlib/testatexit
+ stdlib/test-canon
+ stdlib/test-canon2
++stdlib/test-mkostemp-O_CLOEXEC
++stdlib/test-mkostemp-child
+ stdlib/teston_exit
+ stdlib/teststrtol
+ stdlib/teststrtoq
+diff --git a/test/stdlib/test-mkostemp-O_CLOEXEC.c b/test/stdlib/test-mkostemp-O_CLOEXEC.c
+new file mode 100644
+index 0000000..9ff229a
+--- /dev/null
++++ b/test/stdlib/test-mkostemp-O_CLOEXEC.c
+@@ -0,0 +1,45 @@
++#define _XOPEN_SOURCE_EXTENDED
++#include <stdio.h>
++#include <stdlib.h>
++#include <string.h>
++#include <unistd.h>
++#include <sys/types.h>
++#include <sys/stat.h>
++#include <fcntl.h>
++#include <sys/wait.h>
++#include <errno.h>
++
++#if !defined __ARCH_USE_MMU__
++# define fork vfork
++#endif
++
++int main(int argc, char *argv[]) {
++    int fd, status;
++    char buff[5];
++    char template[] = "/tmp/test-mkostemp.XXXXXX";
++
++    fd = mkostemp(template, O_CLOEXEC);
++    unlink(template);
++
++    snprintf(buff, 5, "%d", fd);
++
++    if(!fork())
++        if(execl("./test-mkostemp-child", "test-mkostemp-child", buff, NULL) == -1)
++            exit(EXIT_FAILURE);
++
++    wait(&status);
++
++    memset(buff, 0, 5);
++    lseek(fd, 0, SEEK_SET);
++    errno = 0;
++    if(read(fd, buff, 5) == -1)
++        exit(EXIT_FAILURE);
++
++    if(!strncmp(buff, "test", 5))
++        exit(EXIT_FAILURE);
++    else
++        exit(EXIT_SUCCESS);
++
++    close(fd);
++    exit(EXIT_SUCCESS);
++}
+diff --git a/test/stdlib/test-mkostemp-child.c b/test/stdlib/test-mkostemp-child.c
+new file mode 100644
+index 0000000..708bd80
+--- /dev/null
++++ b/test/stdlib/test-mkostemp-child.c
+@@ -0,0 +1,22 @@
++#include <stdio.h>
++#include <stdlib.h>
++#include <unistd.h>
++
++int main(int argc, char *argv[]) {
++    int fd;
++
++    /* This file gets built and run as a test, but its
++     * really just a helper for test-mkostemp-O_CLOEXEC.c.
++     * So, we'll always return succcess.
++     */
++    if(argc != 2)
++        exit(EXIT_SUCCESS);
++
++    sscanf(argv[1], "%d", &fd);
++
++    if(write(fd, "test\0", 5) == -1)
++        ; /* Don't Panic!  Failure is okay here. */
++
++    close(fd);
++    exit(EXIT_SUCCESS);
++}
+-- 
+1.9.1
+
diff --git a/patches/uClibc/0.9.33.2/401-libc-stdlib-add-mkostemp-helpers.patch b/patches/uClibc/0.9.33.2/401-libc-stdlib-add-mkostemp-helpers.patch
new file mode 100644
index 0000000..1a8d7a5
--- /dev/null
+++ b/patches/uClibc/0.9.33.2/401-libc-stdlib-add-mkostemp-helpers.patch
@@ -0,0 +1,155 @@
+From 42d1b23fc0f3748b8bf474e456d6c44aa7e563fd Mon Sep 17 00:00:00 2001
+From: Mike Frysinger <vapier@gentoo.org>
+Date: Wed, 14 Nov 2012 00:30:54 -0500
+Subject: [PATCH 1/1] libc/stdlib: add mkostemp helpers
+
+Some projects (like udev) are starting to use this.
+
+Imported from glibc.
+
+Signed-off-by: Mike Frysinger <vapier@gentoo.org>
+---
+ include/stdlib.h         | 23 +++++++++++++++++++++++
+ libc/stdlib/Makefile.in  |  4 ++--
+ libc/stdlib/mkostemp.c   | 32 ++++++++++++++++++++++++++++++++
+ libc/stdlib/mkostemp64.c | 33 +++++++++++++++++++++++++++++++++
+ 4 files changed, 90 insertions(+), 2 deletions(-)
+ create mode 100644 libc/stdlib/mkostemp.c
+ create mode 100644 libc/stdlib/mkostemp64.c
+
+diff --git a/include/stdlib.h b/include/stdlib.h
+index 354fc66..79ccc55 100644
+--- a/include/stdlib.h
++++ b/include/stdlib.h
+@@ -652,6 +652,29 @@ extern int mkstemp64 (char *__template) __nonnull ((1)) __wur;
+ extern char *mkdtemp (char *__template) __THROW __nonnull ((1)) __wur;
+ #endif
+ 
++#ifdef __USE_GNU
++/* Generate a unique temporary file name from TEMPLATE similar to
++   mkstemp.  But allow the caller to pass additional flags which are
++   used in the open call to create the file..
++
++   This function is a possible cancellation point and therefore not
++   marked with __THROW.  */
++# ifndef __USE_FILE_OFFSET64
++extern int mkostemp (char *__template, int __flags) __nonnull ((1)) __wur;
++# else
++#  ifdef __REDIRECT
++extern int __REDIRECT (mkostemp, (char *__template, int __flags), mkostemp64)
++     __nonnull ((1)) __wur;
++#  else
++#   define mkostemp mkostemp64
++#  endif
++# endif
++# ifdef __USE_LARGEFILE64
++extern int mkostemp64 (char *__template, int __flags) __nonnull ((1)) __wur;
++# endif
++
++#endif
++
+ 
+ __BEGIN_NAMESPACE_STD
+ /* Execute the given line as a shell command.
+diff --git a/libc/stdlib/Makefile.in b/libc/stdlib/Makefile.in
+index 3166b8e..b92f7ce 100644
+--- a/libc/stdlib/Makefile.in
++++ b/libc/stdlib/Makefile.in
+@@ -12,7 +12,7 @@ include $(top_srcdir)libc/stdlib/malloc-simple/Makefile.in
+ include $(top_srcdir)libc/stdlib/malloc-standard/Makefile.in
+ 
+ CSRC-y := \
+-	abort.c getenv.c mkdtemp.c realpath.c canonicalize.c mkstemp.c \
++	abort.c getenv.c mkdtemp.c realpath.c canonicalize.c mkstemp.c mkostemp.c \
+ 	rand.c random.c random_r.c setenv.c div.c ldiv.c lldiv.c \
+ 	getpt.c drand48-iter.c jrand48.c \
+ 	jrand48_r.c lcong48.c lrand48.c lrand48_r.c mrand48.c mrand48_r.c nrand48.c \
+@@ -21,7 +21,7 @@ CSRC-y := \
+ CSRC-$(UCLIBC_HAS_ADVANCED_REALTIME) += posix_memalign.c
+ CSRC-$(UCLIBC_HAS_PTY) += grantpt.c unlockpt.c ptsname.c
+ CSRC-$(UCLIBC_HAS_ARC4RANDOM) += arc4random.c
+-CSRC-$(UCLIBC_HAS_LFS) += mkstemp64.c
++CSRC-$(UCLIBC_HAS_LFS) += mkstemp64.c mkostemp64.c
+ CSRC-$(UCLIBC_HAS_FLOATS) += drand48.c drand48_r.c erand48.c erand48_r.c
+ CSRC-$(if $(findstring yy,$(UCLIBC_HAS_FLOATS)$(UCLIBC_SUSV3_LEGACY)),y) += \
+ 	gcvt.c
+diff --git a/libc/stdlib/mkostemp.c b/libc/stdlib/mkostemp.c
+new file mode 100644
+index 0000000..93b50fc
+--- /dev/null
++++ b/libc/stdlib/mkostemp.c
+@@ -0,0 +1,32 @@
++/* Copyright (C) 1998-2012 Free Software Foundation, Inc.
++   This file is part of the GNU C Library.
++
++   The GNU C Library is free software; you can redistribute it and/or
++   modify it under the terms of the GNU Lesser General Public
++   License as published by the Free Software Foundation; either
++   version 2.1 of the License, or (at your option) any later version.
++
++   The GNU C Library is distributed in the hope that it will be useful,
++   but WITHOUT ANY WARRANTY; without even the implied warranty of
++   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
++   Lesser General Public License for more details.
++
++   You should have received a copy of the GNU Lesser General Public
++   License along with the GNU C Library; if not, see
++   <http://www.gnu.org/licenses/>.  */
++
++#include <stdio.h>
++#include <stdlib.h>
++#include "../misc/internals/tempname.h"
++
++/* Generate a unique temporary file name from TEMPLATE.
++   The last six characters of TEMPLATE must be "XXXXXX";
++   they are replaced with a string that makes the filename unique.
++   Then open the file and return a fd. */
++int
++mkostemp (template, flags)
++     char *template;
++     int flags;
++{
++  return __gen_tempname (template, __GT_FILE, flags);
++}
+diff --git a/libc/stdlib/mkostemp64.c b/libc/stdlib/mkostemp64.c
+new file mode 100644
+index 0000000..5509d8c
+--- /dev/null
++++ b/libc/stdlib/mkostemp64.c
+@@ -0,0 +1,33 @@
++/* Copyright (C) 2000-2012 Free Software Foundation, Inc.
++   This file is part of the GNU C Library.
++
++   The GNU C Library is free software; you can redistribute it and/or
++   modify it under the terms of the GNU Lesser General Public
++   License as published by the Free Software Foundation; either
++   version 2.1 of the License, or (at your option) any later version.
++
++   The GNU C Library is distributed in the hope that it will be useful,
++   but WITHOUT ANY WARRANTY; without even the implied warranty of
++   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
++   Lesser General Public License for more details.
++
++   You should have received a copy of the GNU Lesser General Public
++   License along with the GNU C Library; if not, see
++   <http://www.gnu.org/licenses/>.  */
++
++#include <fcntl.h>
++#include <stdio.h>
++#include <stdlib.h>
++#include "../misc/internals/tempname.h"
++
++/* Generate a unique temporary file name from TEMPLATE.
++   The last six characters of TEMPLATE must be "XXXXXX";
++   they are replaced with a string that makes the filename unique.
++   Then open the file and return a fd. */
++int
++mkostemp64 (template, flags)
++     char *template;
++     int flags;
++{
++  return __gen_tempname (template, __GT_BIGFILE, flags | O_LARGEFILE);
++}
+-- 
+1.9.1
+
diff --git a/patches/uClibc/0.9.33.2/402-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch b/patches/uClibc/0.9.33.2/402-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch
new file mode 100644
index 0000000..b086be2
--- /dev/null
+++ b/patches/uClibc/0.9.33.2/402-stdlib-Fix-fndecl-of-mkostemp-mkostemp64.patch
@@ -0,0 +1,44 @@
+From 5084cf8b6b8f01890b62bfd0edf085d04887d084 Mon Sep 17 00:00:00 2001
+From: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
+Date: Sun, 10 Feb 2013 17:21:15 +0100
+Subject: [PATCH 1/1] stdlib: Fix fndecl of mkostemp, mkostemp64
+
+Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
+---
+ libc/stdlib/mkostemp.c   | 4 +---
+ libc/stdlib/mkostemp64.c | 4 +---
+ 2 files changed, 2 insertions(+), 6 deletions(-)
+
+diff --git a/libc/stdlib/mkostemp.c b/libc/stdlib/mkostemp.c
+index 93b50fc..0369235 100644
+--- a/libc/stdlib/mkostemp.c
++++ b/libc/stdlib/mkostemp.c
+@@ -24,9 +24,7 @@
+    they are replaced with a string that makes the filename unique.
+    Then open the file and return a fd. */
+ int
+-mkostemp (template, flags)
+-     char *template;
+-     int flags;
++mkostemp (char *template, int flags)
+ {
+   return __gen_tempname (template, __GT_FILE, flags);
+ }
+diff --git a/libc/stdlib/mkostemp64.c b/libc/stdlib/mkostemp64.c
+index 5509d8c..d21def5 100644
+--- a/libc/stdlib/mkostemp64.c
++++ b/libc/stdlib/mkostemp64.c
+@@ -25,9 +25,7 @@
+    they are replaced with a string that makes the filename unique.
+    Then open the file and return a fd. */
+ int
+-mkostemp64 (template, flags)
+-     char *template;
+-     int flags;
++mkostemp64 (char *template, int flags)
+ {
+   return __gen_tempname (template, __GT_BIGFILE, flags | O_LARGEFILE);
+ }
+-- 
+1.9.1
+
diff --git a/patches/uClibc/0.9.33.2/403-mkostemp-fix-implementation.patch b/patches/uClibc/0.9.33.2/403-mkostemp-fix-implementation.patch
new file mode 100644
index 0000000..79b9432
--- /dev/null
+++ b/patches/uClibc/0.9.33.2/403-mkostemp-fix-implementation.patch
@@ -0,0 +1,328 @@
+From 638a23483b40c5b606ee323e6612e7e454e5154b Mon Sep 17 00:00:00 2001
+From: "Anthony G. Basile" <blueness@gentoo.org>
+Date: Mon, 27 Oct 2014 16:13:34 -0400
+Subject: [PATCH 1/1] mkostemp: fix implementation
+
+mkostemp(char *template, int flags) generates a unique temporary
+filename from a template.  The flags parameter accepts three of
+the same flags as open(2): O_APPEND, O_CLOEXEC, and O_SYNC.  The
+current implementation of mkostemp(3) does not respect the flags
+and in fact confuses the flags with the file mode which should
+always be S_IRUSR | S_IWUSR.  This patch corrects this issue.
+
+Signed-off-by: Anthony G. Basile <blueness@gentoo.org>
+Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
+---
+ libc/inet/getaddrinfo.c               |  2 +-
+ libc/misc/internals/tempname.c        |  6 ++---
+ libc/misc/internals/tempname.h        |  2 +-
+ libc/stdio/tempnam.c                  |  2 +-
+ libc/stdio/tmpfile.c                  |  2 +-
+ libc/stdio/tmpnam.c                   |  2 +-
+ libc/stdio/tmpnam_r.c                 |  2 +-
+ libc/stdlib/mkdtemp.c                 |  2 +-
+ libc/stdlib/mkostemp.c                |  4 +++-
+ libc/stdlib/mkostemp64.c              |  2 +-
+ libc/stdlib/mkstemp.c                 |  2 +-
+ libc/stdlib/mkstemp64.c               |  2 +-
+ libc/stdlib/mktemp.c                  |  2 +-
+ libpthread/nptl/sem_open.c            |  2 +-
+ test/.gitignore                       |  2 ++
+ test/stdlib/test-mkostemp-O_CLOEXEC.c | 45 +++++++++++++++++++++++++++++++++++
+ test/stdlib/test-mkostemp-child.c     | 22 +++++++++++++++++
+ 17 files changed, 87 insertions(+), 16 deletions(-)
+ create mode 100644 test/stdlib/test-mkostemp-O_CLOEXEC.c
+ create mode 100644 test/stdlib/test-mkostemp-child.c
+
+diff --git a/libc/inet/getaddrinfo.c b/libc/inet/getaddrinfo.c
+index b61d69c..168adb1 100644
+--- a/libc/inet/getaddrinfo.c
++++ b/libc/inet/getaddrinfo.c
+@@ -308,7 +308,7 @@ gaih_local(const char *name, const struct gaih_service *service,
+ 		char *buf = ((struct sockaddr_un *)ai->ai_addr)->sun_path;
+ 
+ 		if (__path_search(buf, L_tmpnam, NULL, NULL, 0) != 0
+-		 || __gen_tempname(buf, __GT_NOCREATE, 0) != 0
++		 || __gen_tempname(buf, __GT_NOCREATE, 0, 0) != 0
+ 		) {
+ 			return -EAI_SYSTEM;
+ 		}
+diff --git a/libc/misc/internals/tempname.c b/libc/misc/internals/tempname.c
+index 18fd823..edcc31c 100644
+--- a/libc/misc/internals/tempname.c
++++ b/libc/misc/internals/tempname.c
+@@ -177,7 +177,7 @@ static void brain_damaged_fillrand(unsigned char *buf, unsigned int len)
+    __GT_DIR:            create a directory with given mode.
+ 
+ */
+-int attribute_hidden __gen_tempname (char *tmpl, int kind, mode_t mode)
++int attribute_hidden __gen_tempname (char *tmpl, int kind, int flags, mode_t mode)
+ {
+     char *XXXXXX;
+     unsigned int i;
+@@ -219,11 +219,11 @@ int attribute_hidden __gen_tempname (char *tmpl, int kind, mode_t mode)
+ 			fd = 0;
+ 		}
+ 	    case __GT_FILE:
+-		fd = open (tmpl, O_RDWR | O_CREAT | O_EXCL, mode);
++		fd = open (tmpl, O_RDWR | O_CREAT | O_EXCL | flags, mode);
+ 		break;
+ #if defined __UCLIBC_HAS_LFS__
+ 	    case __GT_BIGFILE:
+-		fd = open64 (tmpl, O_RDWR | O_CREAT | O_EXCL, mode);
++		fd = open64 (tmpl, O_RDWR | O_CREAT | O_EXCL | flags, mode);
+ 		break;
+ #endif
+ 	    case __GT_DIR:
+diff --git a/libc/misc/internals/tempname.h b/libc/misc/internals/tempname.h
+index e75b632..edfe26d 100644
+--- a/libc/misc/internals/tempname.h
++++ b/libc/misc/internals/tempname.h
+@@ -10,7 +10,7 @@ extern int ___path_search (char *tmpl, size_t tmpl_len, const char *dir,
+ 	        const char *pfx /*, int try_tmpdir */) attribute_hidden;
+ #define __path_search(tmpl, tmpl_len, dir, pfx, try_tmpdir) ___path_search(tmpl, tmpl_len, dir, pfx)
+ 
+-extern int __gen_tempname (char *__tmpl, int __kind, mode_t mode) attribute_hidden;
++extern int __gen_tempname (char *__tmpl, int __kind, int flags, mode_t mode) attribute_hidden;
+ 
+ /* The __kind argument to __gen_tempname may be one of: */
+ #define __GT_FILE     0       /* create a file */
+diff --git a/libc/stdio/tempnam.c b/libc/stdio/tempnam.c
+index 232ed02..74bb26e 100644
+--- a/libc/stdio/tempnam.c
++++ b/libc/stdio/tempnam.c
+@@ -35,7 +35,7 @@ tempnam (const char *dir, const char *pfx)
+   if (__path_search (buf, FILENAME_MAX, dir, pfx, 1))
+     return NULL;
+ 
+-  if (__gen_tempname (buf, __GT_NOCREATE, 0))
++  if (__gen_tempname (buf, __GT_NOCREATE, 0, 0))
+     return NULL;
+ 
+   return strdup (buf);
+diff --git a/libc/stdio/tmpfile.c b/libc/stdio/tmpfile.c
+index a9bf474..83c85b5 100644
+--- a/libc/stdio/tmpfile.c
++++ b/libc/stdio/tmpfile.c
+@@ -35,7 +35,7 @@ FILE * tmpfile (void)
+ 
+     if (__path_search (buf, FILENAME_MAX, NULL, "tmpf", 0))
+ 	return NULL;
+-    fd = __gen_tempname (buf, __GT_FILE, S_IRUSR | S_IWUSR);
++    fd = __gen_tempname (buf, __GT_FILE, 0, S_IRUSR | S_IWUSR);
+     if (fd < 0)
+ 	return NULL;
+ 
+diff --git a/libc/stdio/tmpnam.c b/libc/stdio/tmpnam.c
+index 88b9bff..ffed862 100644
+--- a/libc/stdio/tmpnam.c
++++ b/libc/stdio/tmpnam.c
+@@ -40,7 +40,7 @@ tmpnam (char *s)
+ 			0))
+     return NULL;
+ 
+-  if (__builtin_expect (__gen_tempname (tmpbuf, __GT_NOCREATE, 0), 0))
++  if (__builtin_expect (__gen_tempname (tmpbuf, __GT_NOCREATE, 0, 0), 0))
+     return NULL;
+ 
+   if (s == NULL)
+diff --git a/libc/stdio/tmpnam_r.c b/libc/stdio/tmpnam_r.c
+index 8cde84d..bfd60a4 100644
+--- a/libc/stdio/tmpnam_r.c
++++ b/libc/stdio/tmpnam_r.c
+@@ -27,7 +27,7 @@ char * tmpnam_r (char *s)
+ 
+     if (__path_search (s, L_tmpnam, NULL, NULL, 0))
+ 	return NULL;
+-    if (__gen_tempname (s, __GT_NOCREATE, 0))
++    if (__gen_tempname (s, __GT_NOCREATE, 0, 0))
+ 	return NULL;
+ 
+     return s;
+diff --git a/libc/stdlib/mkdtemp.c b/libc/stdlib/mkdtemp.c
+index da7598a..e6d4a36 100644
+--- a/libc/stdlib/mkdtemp.c
++++ b/libc/stdlib/mkdtemp.c
+@@ -29,7 +29,7 @@
+    (This function comes from OpenBSD.) */
+ char * mkdtemp (char *template)
+ {
+-  if (__gen_tempname (template, __GT_DIR, S_IRUSR | S_IWUSR | S_IXUSR))
++  if (__gen_tempname (template, __GT_DIR, 0, S_IRUSR | S_IWUSR | S_IXUSR))
+     return NULL;
+   else
+     return template;
+diff --git a/libc/stdlib/mkostemp.c b/libc/stdlib/mkostemp.c
+index 0369235..912be30 100644
+--- a/libc/stdlib/mkostemp.c
++++ b/libc/stdlib/mkostemp.c
+@@ -17,6 +17,7 @@
+ 
+ #include <stdio.h>
+ #include <stdlib.h>
++#include <fcntl.h>
+ #include "../misc/internals/tempname.h"
+ 
+ /* Generate a unique temporary file name from TEMPLATE.
+@@ -26,5 +27,6 @@
+ int
+ mkostemp (char *template, int flags)
+ {
+-  return __gen_tempname (template, __GT_FILE, flags);
++  flags -= flags & O_ACCMODE; /* Remove O_RDONLY, O_WRONLY, and O_RDWR. */
++  return __gen_tempname (template, __GT_FILE, flags, S_IRUSR | S_IWUSR);
+ }
+diff --git a/libc/stdlib/mkostemp64.c b/libc/stdlib/mkostemp64.c
+index d21def5..c6d6d84 100644
+--- a/libc/stdlib/mkostemp64.c
++++ b/libc/stdlib/mkostemp64.c
+@@ -27,5 +27,5 @@
+ int
+ mkostemp64 (char *template, int flags)
+ {
+-  return __gen_tempname (template, __GT_BIGFILE, flags | O_LARGEFILE);
++  return __gen_tempname (template, __GT_BIGFILE, flags | O_LARGEFILE, S_IRUSR | S_IWUSR | S_IXUSR);
+ }
+diff --git a/libc/stdlib/mkstemp.c b/libc/stdlib/mkstemp.c
+index 61c7175..a3a1595 100644
+--- a/libc/stdlib/mkstemp.c
++++ b/libc/stdlib/mkstemp.c
+@@ -26,5 +26,5 @@
+    Then open the file and return a fd. */
+ int mkstemp (char *template)
+ {
+-    return __gen_tempname (template, __GT_FILE, S_IRUSR | S_IWUSR);
++    return __gen_tempname (template, __GT_FILE, 0, S_IRUSR | S_IWUSR);
+ }
+diff --git a/libc/stdlib/mkstemp64.c b/libc/stdlib/mkstemp64.c
+index e29be2d..6f2ee3e 100644
+--- a/libc/stdlib/mkstemp64.c
++++ b/libc/stdlib/mkstemp64.c
+@@ -26,5 +26,5 @@
+    Then open the file and return a fd. */
+ int mkstemp64 (char *template)
+ {
+-    return __gen_tempname (template, __GT_BIGFILE, S_IRUSR | S_IWUSR);
++    return __gen_tempname (template, __GT_BIGFILE, 0, S_IRUSR | S_IWUSR);
+ }
+diff --git a/libc/stdlib/mktemp.c b/libc/stdlib/mktemp.c
+index edd001d..1ff93da 100644
+--- a/libc/stdlib/mktemp.c
++++ b/libc/stdlib/mktemp.c
+@@ -24,7 +24,7 @@
+  * they are replaced with a string that makes the filename unique.  */
+ char *mktemp(char *template)
+ {
+-	if (__gen_tempname (template, __GT_NOCREATE, 0) < 0)
++	if (__gen_tempname (template, __GT_NOCREATE, 0, 0) < 0)
+ 		/* We return the null string if we can't find a unique file name.  */
+ 		template[0] = '\0';
+ 
+diff --git a/libpthread/nptl/sem_open.c b/libpthread/nptl/sem_open.c
+index 1b36164..3a72079 100644
+--- a/libpthread/nptl/sem_open.c
++++ b/libpthread/nptl/sem_open.c
+@@ -336,7 +336,7 @@ sem_open (const char *name, int oflag, ...)
+       mempcpy (mempcpy (tmpfname, mountpoint.dir, mountpoint.dirlen),
+ 	"XXXXXX", 7);
+ 
+-      fd = __gen_tempname (tmpfname, __GT_FILE, mode);
++      fd = __gen_tempname (tmpfname, __GT_FILE, 0, mode);
+       if (fd == -1)
+         return SEM_FAILED;
+ 
+diff --git a/test/.gitignore b/test/.gitignore
+index 8f32031..5944f0a 100644
+--- a/test/.gitignore
++++ b/test/.gitignore
+@@ -274,6 +274,8 @@ stdlib/testarc4random
+ stdlib/testatexit
+ stdlib/test-canon
+ stdlib/test-canon2
++stdlib/test-mkostemp-O_CLOEXEC
++stdlib/test-mkostemp-child
+ stdlib/teston_exit
+ stdlib/teststrtol
+ stdlib/teststrtoq
+diff --git a/test/stdlib/test-mkostemp-O_CLOEXEC.c b/test/stdlib/test-mkostemp-O_CLOEXEC.c
+new file mode 100644
+index 0000000..9ff229a
+--- /dev/null
++++ b/test/stdlib/test-mkostemp-O_CLOEXEC.c
+@@ -0,0 +1,45 @@
++#define _XOPEN_SOURCE_EXTENDED
++#include <stdio.h>
++#include <stdlib.h>
++#include <string.h>
++#include <unistd.h>
++#include <sys/types.h>
++#include <sys/stat.h>
++#include <fcntl.h>
++#include <sys/wait.h>
++#include <errno.h>
++
++#if !defined __ARCH_USE_MMU__
++# define fork vfork
++#endif
++
++int main(int argc, char *argv[]) {
++    int fd, status;
++    char buff[5];
++    char template[] = "/tmp/test-mkostemp.XXXXXX";
++
++    fd = mkostemp(template, O_CLOEXEC);
++    unlink(template);
++
++    snprintf(buff, 5, "%d", fd);
++
++    if(!fork())
++        if(execl("./test-mkostemp-child", "test-mkostemp-child", buff, NULL) == -1)
++            exit(EXIT_FAILURE);
++
++    wait(&status);
++
++    memset(buff, 0, 5);
++    lseek(fd, 0, SEEK_SET);
++    errno = 0;
++    if(read(fd, buff, 5) == -1)
++        exit(EXIT_FAILURE);
++
++    if(!strncmp(buff, "test", 5))
++        exit(EXIT_FAILURE);
++    else
++        exit(EXIT_SUCCESS);
++
++    close(fd);
++    exit(EXIT_SUCCESS);
++}
+diff --git a/test/stdlib/test-mkostemp-child.c b/test/stdlib/test-mkostemp-child.c
+new file mode 100644
+index 0000000..708bd80
+--- /dev/null
++++ b/test/stdlib/test-mkostemp-child.c
+@@ -0,0 +1,22 @@
++#include <stdio.h>
++#include <stdlib.h>
++#include <unistd.h>
++
++int main(int argc, char *argv[]) {
++    int fd;
++
++    /* This file gets built and run as a test, but its
++     * really just a helper for test-mkostemp-O_CLOEXEC.c.
++     * So, we'll always return succcess.
++     */
++    if(argc != 2)
++        exit(EXIT_SUCCESS);
++
++    sscanf(argv[1], "%d", &fd);
++
++    if(write(fd, "test\0", 5) == -1)
++        ; /* Don't Panic!  Failure is okay here. */
++
++    close(fd);
++    exit(EXIT_SUCCESS);
++}
+-- 
+1.9.1
+
